examples/self_sustaining_parallel.py

```diff
@@ -0,0 +1,205 @@
+# The following class implements the computational model. It can be easily
+# adapted to various usecases, since the functions that define the initial
+# tasks and the feedback loop are abstracted
+
+import joblib
+from queue import SimpleQueue
+
+
+class SelfSustainingParallelWork:
+
+    def __init__(self, parallel_n_jobs, parallel_batch_size=1):
+        self.parallel = joblib.Parallel(
+            n_jobs=parallel_n_jobs,
+            return_as="generator_unordered",
+            batch_size=parallel_batch_size,
+        )
+
+        if self.parallel.batch_size == "auto":
+            # Adaptive batch sizing might set batch size to a value higher than
+            # `n_jobs` during compute, which would cause a deadlock, since
+            # then `joblib.Parallel` would risk having to wait for more jobs
+            # than the feedback loop can provide.
+            raise ValueError(
+                "Setting joblib.Parallel `batch_size` parameter to 'auto' is "
+                "not allowed in SelfSustainingParallelWork"
+            )
+
+    def __call__(
+            self,
+            starting_tasks,
+            task_from_output_fn,
+            task_completion_callback
+    ):
+        """Main function that starts the compute and the feedback loop.
+
+        Parameters
+        ----------
+        starting_tasks: iterable of functions
+            An iterable of functions that take no arguments. The compute will
+            start by dispatching those tasks to the workers.
+
+        task_from_output_fn: function
+            A function that the feedback loop will used to create new tasks
+            from previous results. The function must expect one argument, which
+            will be passed the output returned by the latest completed task. It
+            must return either a joblib.delayed-like function, either None, in
+            which case no new task will be dispatched.
+
+        task_completion_callback: function
+            A function that will be called in the calling thread each time an
+            output is retrieved, taking said output as argument. It can be used
+            for recording, plotting, logging,...
+        """
+        # It's required to use a multi-threaded queue here, so that the
+        # callback threads in joblib.Parallel internal mechanisms can also
+        # get the new tasks as soon as it's available and dispatch it.
+        task_feedback_queue = SimpleQueue()
+
+        # Start the compute...
+        input_generator = self._input_generator(
+            starting_tasks, task_from_output_fn, task_feedback_queue
+        )
+        _output_generator = self.parallel(input_generator)
+
+        # ... start retrieving outputs...
+        for output in _output_generator:
+            task_completion_callback(output)
+
+            # ...create a new task based on the outputs...
+            try:
+                new_task = task_from_output_fn(output)
+            except BaseException:
```

Comment:
If we make this a first-class class of joblib, maybe we should support stopping the iteration via a "raise StopIteration" and thus capture it here. It is a somewhat canonical way of stopping an iteration (though I suspect that many do not know it)

Reply:
The stopping criterion is actually expected to be triggered by returning `None` rather than a `joblib.delayed`-like task in `task_from_output_fn`, so the current purpose from this `try/except` block is not to allow the user to stop the compute by raising an exception. The purpose is to avoid a bug that deadlocks the backend because the input iterator would not be closed if an exception is raised in `__call__`.

Two follow-up point from this:
- maybe stopping the compute by returning `None` in `task_from_output_fn` is not so good design :thought_balloon: 
- the whole loop should be wrapped in this try/except block to also catch `KeyboardInterrupt`. 

Like in a previous conversation, let's iterate to refine the api.